cache/cache.go

```diff
@@ -0,0 +1,43 @@
+package cache // import "bosun.org/cache"
+
+import (
+	"bosun.org/_third_party/github.com/golang/groupcache/lru"
+	"bosun.org/_third_party/github.com/golang/groupcache/singleflight"
+	"sync"
+)
+
+type Cache struct {
+	g singleflight.Group
+
+	sync.Mutex
+	lru *lru.Cache
+}
+
+type Result struct {
+	Val interface{}
+	Err error
+}
+
+func New() *Cache {
+	return &Cache{
+		lru: lru.New(100),
+	}
+}
+
+func (c *Cache) Get(key string, getFn func() (interface{}, error)) (interface{}, error) {
+	c.Lock()
+	result, ok := c.lru.Get(key)
+	c.Unlock()
+	if ok {
+		res := result.(Result)
+		return res.Val, res.Err
+	}
+	// our lock only serves to protect the lru.
+	// we can (and should!) do singleflight requests concurently
+	val, err := c.g.Do(key, getFn)
+	r := Result{val, err}
+	c.Lock()
+	c.lru.Add(key, r)
```

Comment:
in practice i don't think this would be an issue, and in current context the lock contention will be minimal.
however your approach is more correct and allows the cache to be used in high-request scenarios, so i'll adjust.

frankly it's a little messy that the lock is on global level, so we release lock after checking (we don't want to hold a global lock during the getFn() call) but this also allows for race conditions. (object appearing in cache right after we checked, and hence we still call getFn() needlessly).  not an issue in the current context, but could be done more correctly